Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +636,490 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        addStringifyOutcome(reason);
+#endif
+    m_length = std::size(m_buffer) + 1;
+}
+
+#if FAST_STRINGIFY_HISTOGRAM
+
+inline void FastStringifier::recordFailure(String&& reason)
+{
+    if (!haveFailure())
+        addStringifyOutcome(WTFMove(reason));
+    m_length = std::size(m_buffer) + 1;
+}
+
+#endif
+
+inline void FastStringifier::recordBufferFull()
+{
+    recordFailure("buffer full"_s);
+}
+
+inline unsigned FastStringifier::remainingCapacity() const
+{
+    ASSERT(!haveFailure());
+    return std::size(m_buffer) - m_length;
+}
+
+static bool mayNeedJSONEscaping(LChar character)
+{
+    // FIXME: Use an array like escapedFormsForJSON for speed.
+    return character < ' ' || character == '"' || character == '\\';
+}
+
+#if FAST_STRINGIFY_HISTOGRAM
+
+static String firstGetterSetterPropertyName(JSGlobalObject& globalObject, JSObject& object)
+{
+    VM& vm = globalObject.vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    PropertyNameArray names(globalObject.vm(), PropertyNameMode::Strings, PrivateSymbolMode::Include);
+    JSObject::getOwnPropertyNames(&object, &globalObject, names, DontEnumPropertiesMode::Include);
+    CLEAR_AND_RETURN_IF_EXCEPTION(scope, "getOwnPropertyNames exception occurred"_s);
+    for (auto& name : names) {
+        PropertySlot slot(&object, PropertySlot::InternalMethodType::Get);
+        JSObject::getOwnPropertySlot(&object, &globalObject, name, slot);
+        CLEAR_AND_RETURN_IF_EXCEPTION(scope, "getOwnPropertySlot exception occurred"_s);
+        if (slot.isAccessor())
+            return name.string();
+    }
+    return "not found"_s;
+}
+
+#endif
+
+inline bool FastStringifier::mayHaveToJSON(JSObject& object) const
+{
+    // FIXME: Add test cases to show what happens where toJSON has a getter with side effects; likely this handles that edge case incorrectly.
+    // If we decide this is not OK, we can consider this alternate approach:
+    // - Check only the object and array prototypes to see if they have toJSON, perhaps lazily when we encounter the first object or array.
+    //   Would be good to cache this too, but unclear exactly how to do that.
+    // - For objects, rely on the fact that a callable toJSON property is not serializable, and so such a property on the object, not the
+    //   prototype, will cause the fast stringifier to fail when we hit it later, so no check of the object is needed.
+    // - May need to add a check that each object/array we serialize has the standard prototype by checking hasMonoProto then comparing
+    //   storedPrototype with objectPrototype().
+    // The alternate approach may be faster since we will not need to call cachedSpecialProperty for each object.
+    auto function = object.structure()->cachedSpecialProperty(CachedSpecialPropertyKey::ToJSON);
+    if (!function) {
+        auto scope = DECLARE_THROW_SCOPE(m_vm);
+        scope.assertNoException();
+        PropertySlot slot(&object, PropertySlot::InternalMethodType::Get);
```

Comment:
I don't think using PropertySlot::InternalMethodType::Get is side effect free.  We normally use PropertySlot::InternalMethodType::VMInquiry when we don't want side effects.  However, I think VMInquiry will result in getPropertySlot returning false if there's a potential for side effects, which is the opposite of what you want.

@Constellation can you double check my understanding here?

Reply:
Yes, `PropertySlot::InternalMethodType::Get` is normal `[[Get]]` operation in JS, thus, it can have a side effect (calling getters etc.).

Reply:
Ah, but we are just using `object.getPropertySlot`. So we do not have a side effect.

Reply:
I got rid of the call to getValue.